url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/io/Generic_StaticIO.java

```diff
@@ -0,0 +1,2156 @@
+                        + ".copy(File(" + fileToCopy + "),"
+                        + "File(" + directoryToCopyInto + "))");
+                e.printStackTrace(System.err);
+                System.exit(Generic_ErrorAndExceptionHandler.IOException);
+            }
+        }
+        try {
+            BufferedInputStream bis;
+            bis = getBufferedInputStream(fileToCopy);
+            BufferedOutputStream bos;
+            bos = getBufferedOutputStream(output_File);
+            // bufferSize should be power of 2 (e.g. Math.pow(2, 12)), but nothing too big.
+            int bufferSize = 2048;
+            long numberOfArrayReads = fileToCopy.length() / bufferSize;
+            long numberOfSingleReads = fileToCopy.length() - (numberOfArrayReads * bufferSize);
+            byte[] b = new byte[bufferSize];
+            for (int i = 0; i < numberOfArrayReads; i++) {
+                bis.read(b);
+                bos.write(b);
+            }
+            for (int i = 0; i < numberOfSingleReads; i++) {
+                bos.write(bis.read());
+            }
+            bos.flush();
+            bos.close();
+            bis.close();
+        } catch (IOException e) {
+            System.err.print(e.getMessage());
+            e.printStackTrace(System.err);
+            System.exit(Generic_ErrorAndExceptionHandler.IOException);
+        }
+    }
+
+    /**
+     * @param f
+     * @return <code>new BufferedInputStream(new FileInputStream(f)))</code>
+     */
+    public static BufferedInputStream getBufferedInputStream(File f) {
+        BufferedInputStream result = null;
+        FileInputStream fis;
+        fis = getFileInputStream(f);
+        result = new BufferedInputStream(fis);
+        return result;
+    }
+
+    /**
+     * @param f
+     * @return <code>new FileInputStream(f)</code>
+     */
+    public static FileInputStream getFileInputStream(File f) {
+        FileInputStream result = null;
+        try {
+            result = new FileInputStream(f);
+        } catch (FileNotFoundException ex) {
+            System.err.println("Trying to handle " + ex.getLocalizedMessage());
+            System.err.println("Wait for 2 seconds then trying again to getFileInputStream(File " + f.toString() + ").");
+            if (!f.exists()) {
+                //ex.printStackTrace(System.err);
+                Logger.getLogger(Generic_StaticIO.class.getName()).log(Level.SEVERE, null, ex);
+                // null will be returned...
+            } else {
+                // This can happen because of too many open files.
+                // Try waiting for 2 seconds and then repeating...
+                long wait = 2000L;
+                try {
+                    synchronized (f) {
+                        f.wait(wait);
+                    }
+                } catch (InterruptedException ex2) {
+                    Logger.getLogger(Generic_StaticIO.class.getName()).log(Level.SEVERE, null, ex2);
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 262:268, 299:305, 707:713.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.